app/models/site_setting.rb

```diff
@@ -133,15 +133,20 @@ def self.enable_s3_uploads
     end
 
     def self.absolute_base_url
+      url_basename = SiteSetting.s3_endpoint.split('/')[-1]
       bucket = SiteSetting.enable_s3_uploads ? Discourse.store.s3_bucket_name : GlobalSetting.s3_bucket_name
 
       # cf. http://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region
-      if SiteSetting.Upload.s3_region == "us-east-1"
-        "//#{bucket}.s3.amazonaws.com"
-      elsif SiteSetting.Upload.s3_region == 'cn-north-1'
-        "//#{bucket}.s3.cn-north-1.amazonaws.com.cn"
+      if SiteSetting.s3_endpoint == "https://s3.amazonaws.com"
+        if SiteSetting.Upload.s3_region == "us-east-1"
+          "//#{bucket}.s3.amazonaws.com"
+        elsif SiteSetting.Upload.s3_region == 'cn-north-1'
+          "//#{bucket}.s3.cn-north-1.amazonaws.com.cn"
+        else
+          "//#{bucket}.s3-#{SiteSetting.Upload.s3_region}.amazonaws.com"
+        end
       else
-        "//#{bucket}.s3-#{SiteSetting.Upload.s3_region}.amazonaws.com"
+        "//#{bucket}.#{url_basename}"
```

Comment:
I think this else branch can be removed with 
```
return "//#{bucket}.#{url_basename}" if SiteSetting.s3_endpoint != "https://s3.amazonaws.com"
```

Also, conditions with endpoint can be extracted as methods. And check for cn can be extracted and generalized.

Reply:
Thanks a lot for the review, it's my first ever Ruby app so thanks for the help :grin: 
1) This is what you mean by extracting endpoint as a method?
```
    def self.absolute_base_url
      url_basename = s3_endpoint.split('/')[-1]
    ....
    ....

    def self.s3_endpoint
      SiteSetting.s3_endpoint
    end
```
2) I'm not sure if we have to keep the cn check or not so I'll leave it as it is for now.
3) Is there any advantage of removing the else block other than possibly readability?

Reply:
1. No, I thought about extracting conditions with cn and other regions.
2. If it is here, I think it should be here. Maybe, it should be extended, but its another task.
3. Its decrease of complexity, not readability. Just google refactor if-else statement with return.

Also, according to link http://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region,
there is no need for default URL for `US-EAST1`. It can be removed. And CN endpoints are generated with the same rules with the addition of `.cn` in the end.

So it can be refactored with a simple linear flow like
```
endpoint = "s3.#{SiteSetting.Upload.s3_region}.amazonaws.com"
endpoint << '.cn' if SiteSetting.Upload.s3_region =~ /^cn-*/i
```

Reply:
1,3) I will do that, thanks! :+1: 

Yes, I wanted to change the if-else conditions myself because the conditions seemed unnecessary according to the docs but I didn't refactor it because the code was already written like this earlier. 
I'll change this.

Reply:
Thing is to have no if-else branches. Just linear ifs.
Also, (1) is not needed with the linear flow and remove of default us-east-1.